File: tests/docker-compat/go-client/tests/blocking.go
package tests

import (
	"context"
	"fmt"
	"math"
	"time"

	"github.com/redis/go-redis/v9"
)

func init() {
	Register("blocking", testBlocking)
}

func testBlocking(ctx context.Context, rdb *redis.Client, prefix string) []TestResult {
	var results []TestResult
	cat := "blocking"

	// Helper to create a second connection
	newClient := func() *redis.Client {
		return redis.NewClient(&redis.Options{
			Addr:     rdb.Options().Addr,
			Password: rdb.Options().Password,
			DB:       0,
		})
	}

	// BLPOP_with_data: LPUSH then BLPOP, verify immediate return
	results = append(results, RunTest("BLPOP_with_data", cat, func() error {
		k := Key(prefix, "block:blpop_data")
		defer rdb.Del(ctx, k)

		rdb.RPush(ctx, k, "val1")
		vals, err := rdb.BLPop(ctx, time.Second, k).Result()
		if err != nil {
			return fmt.Errorf("BLPOP failed: %v", err)
		}
		if err := AssertEqual(k, vals[0]); err != nil {
			return err
		}
		return AssertEqual("val1", vals[1])
	}))

	// BLPOP_timeout: BLPOP on empty key with 1s timeout, verify timeout return
	results = append(results, RunTest("BLPOP_timeout", cat, func() error {
		k := Key(prefix, "block:blpop_timeout")
		defer rdb.Del(ctx, k)

		start := time.Now()
		_, err := rdb.BLPop(ctx, time.Second, k).Result()
		elapsed := time.Since(start)

		if err != redis.Nil {
			return fmt.Errorf("expected redis.Nil on timeout, got: %v", err)
		}
		// Should have waited approximately 1 second
		if elapsed < 800*time.Millisecond || elapsed > 2500*time.Millisecond {
			return fmt.Errorf("timeout duration out of range: %v", elapsed)
		}
		return nil
	}))

	// BLPOP_multi_key: BLPOP on multiple keys, LPUSH to second key, verify correct key returned
	results = append(results, RunTest("BLPOP_multi_key", cat, func() error {
		k1 := Key(prefix, "block:multi1")
		k2 := Key(prefix, "block:multi2")
		defer rdb.Del(ctx, k1, k2)

		rdb.RPush(ctx, k2, "from_k2")
		vals, err := rdb.BLPop(ctx, time.Second, k1, k2).Result()
		if err != nil {
			return fmt.Errorf("BLPOP multi_key failed: %v", err)
		}
		if err := AssertEqual(k2, vals[0]); err != nil {
			return err
		}
		return AssertEqual("from_k2", vals[1])
	}))

	// BRPOP_basic: Same as BLPOP but from right side
	results = append(results, RunTest("BRPOP_basic", cat, func() error {
		k := Key(prefix, "block:brpop")
		defer rdb.Del(ctx, k)

		rdb.RPush(ctx, k, "first", "second")
		vals, err := rdb.BRPop(ctx, time.Second, k).Result()
		if err != nil {
			return fmt.Errorf("BRPOP failed: %v", err)
		}
		if err := AssertEqual(k, vals[0]); err != nil {
			return err
		}
		return AssertEqual("second", vals[1])
	}))

	// BLMOVE_basic: BLMOVE from source to dest list, verify move
	results = append(results, RunTest("BLMOVE_basic", cat, func() error {
		src := Key(prefix, "block:blmove_src")
		dst := Key(prefix, "block:blmove_dst")
		defer rdb.Del(ctx, src, dst)

		rdb.RPush(ctx, src, "item1")
		val, err := rdb.BLMove(ctx, src, dst, "LEFT", "RIGHT", time.Second).Result()
		if err != nil {
			return fmt.Errorf("BLMOVE failed: %v", err)
		}
		if err := AssertEqual("item1", val); err != nil {
			return err
		}
		// Verify item is in destination
		dstVals, err := rdb.LRange(ctx, dst, 0, -1).Result()
		if err != nil {
			return fmt.Errorf("LRANGE on dest failed: %v", err)
		}
		return AssertStringSliceEqual([]string{"item1"}, dstVals)
	}))

	// BLMPOP_basic: BLMPOP with LEFT/RIGHT direction
	results = append(results, RunTest("BLMPOP_basic", cat, func() error {
		k := Key(prefix, "block:blmpop")
		defer rdb.Del(ctx, k)

		rdb.RPush(ctx, k, "a", "b", "c")
		// BLMPOP timeout numkeys key ... LEFT|RIGHT
		result, err := rdb.Do(ctx, "BLMPOP", "1", "1", k, "LEFT").Result()
		if err != nil {
			return fmt.Errorf("BLMPOP failed: %v", err)
		}
		// Result should be [key, [elements]]
		arr, ok := result.([]interface{})
		if !ok || len(arr) < 2 {
			return fmt.Errorf("unexpected BLMPOP result format: %v", result)
		}
		retKey := fmt.Sprintf("%v", arr[0])
		if err := AssertEqual(k, retKey); err != nil {
			return err
		}
		// Second element should be a list with the popped element
		elemArr, ok := arr[1].([]interface{})
		if !ok || len(elemArr) < 1 {
			return fmt.Errorf("unexpected BLMPOP element format: %v", arr[1])
		}
		retVal := fmt.Sprintf("%v", elemArr[0])
		return AssertEqual("a", retVal)
	}))

	// BZPOPMIN_basic: BZPOPMIN with data and timeout variants
	results = append(results, RunTest("BZPOPMIN_basic", cat, func() error {
		k := Key(prefix, "block:bzpopmin")
		defer rdb.Del(ctx, k)

		rdb.ZAdd(ctx, k, redis.Z{Score: 1.0, Member: "a"}, redis.Z{Score: 2.0, Member: "b"})
		// Use Do() for reliable result parsing
		result, err := rdb.Do(ctx, "BZPOPMIN", k, "1").Result()
		if err != nil {
			return fmt.Errorf("BZPOPMIN failed: %v", err)
		}
		arr, ok := result.([]interface{})
		if !ok || len(arr) < 3 {
			return fmt.Errorf("unexpected BZPOPMIN result: %v", result)
		}
		retKey := fmt.Sprintf("%v", arr[0])
		retMember := fmt.Sprintf("%v", arr[1])
		retScore := fmt.Sprintf("%v", arr[2])
		if err := AssertEqual(k, retKey); err != nil {
			return err
		}
		if err := AssertEqual("a", retMember); err != nil {
			return err
		}
		return AssertEqual("1", retScore)
	}))

	// BZPOPMAX_basic: BZPOPMAX with data and timeout variants
	results = append(results, RunTest("BZPOPMAX_basic", cat, func() error {
		k := Key(prefix, "block:bzpopmax")
		defer rdb.Del(ctx, k)

		rdb.ZAdd(ctx, k, redis.Z{Score: 1.0, Member: "a"}, redis.Z{Score: 2.0, Member: "b"})
		result, err := rdb.Do(ctx, "BZPOPMAX", k, "1").Result()
		if err != nil {
			return fmt.Errorf("BZPOPMAX failed: %v", err)
		}
		arr, ok := result.([]interface{})
		if !ok || len(arr) < 3 {
			return fmt.Errorf("unexpected BZPOPMAX result: %v", result)
		}
		retKey := fmt.Sprintf("%v", arr[0])
		retMember := fmt.Sprintf("%v", arr[1])
		retScore := fmt.Sprintf("%v", arr[2])
		if err := AssertEqual(k, retKey); err != nil {
			return err
		}
		if err := AssertEqual("b", retMember); err != nil {
			return err
		}
		return AssertEqual("2", retScore)
	}))

	// BLOCKING_concurrent: Producer-consumer pattern
	results = append(results, RunTest("BLOCKING_concurrent", cat, func() error {
		k := Key(prefix, "block:concurrent")
		defer rdb.Del(ctx, k)

		client2 := newClient()
		defer client2.Close()

		// Push data after a delay using goroutine
		done := make(chan error, 1)
		go func() {
			time.Sleep(300 * time.Millisecond)
			done <- client2.RPush(ctx, k, "async_val").Err()
		}()

		// BLPOP should block then receive the pushed value
		vals, err := rdb.BLPop(ctx, 5*time.Second, k).Result()
		if err != nil {
			return fmt.Errorf("BLPOP concurrent failed: %v", err)
		}
		if err := AssertEqual("async_val", vals[1]); err != nil {
			return err
		}
		// Wait for producer goroutine to finish
		if pushErr := <-done; pushErr != nil {
			return fmt.Errorf("producer push failed: %v", pushErr)
		}
		return nil
	}))

	// BLOCKING_timeout_accuracy: Verify timeout duration is within acceptable range
	results = append(results, RunTest("BLOCKING_timeout_accuracy", cat, func() error {
		k := Key(prefix, "block:timeout_accuracy")
		defer rdb.Del(ctx, k)

		start := time.Now()
		_, err := rdb.BLPop(ctx, time.Second, k).Result()
		elapsed := time.Since(start)

		if err != redis.Nil {
			return fmt.Errorf("expected timeout, got: %v", err)
		}
		elapsedMs := float64(elapsed.Milliseconds())
		expectedMs := float64(1000)
		diff := math.Abs(elapsedMs - expectedMs)
		// Allow ±500ms tolerance for CI environments
		if diff > 500 {
			return fmt.Errorf("timeout accuracy out of range: expected ~1000ms, got %.0fms (diff: %.0fms)", elapsedMs, diff)
		}
		return nil
	}))

	return results
}

File: tests/docker-compat/js-client/tests/blocking.js
const Redis = require('ioredis');
const { register, runTest, key, assertEqual, assertTrue, sleep } = require('./framework');

register('blocking', async (redis, prefix) => {
  const results = [];
  const cat = 'blocking';

  // Helper to create a second connection
  const createClient = () => {
    return new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
      retryStrategy: () => null,
      maxRetriesPerRequest: 3,
    });
  };

  // BLPOP_with_data: LPUSH then BLPOP, verify immediate return
  results.push(await runTest('BLPOP_with_data', cat, async () => {
    const k = key(prefix, 'block:blpop_data');
    try {
      await redis.rpush(k, 'val1');
      const result = await redis.blpop(k, 1);
      assertEqual(k, result[0]);
      assertEqual('val1', result[1]);
    } finally {
      await redis.del(k);
    }
  }));

  // BLPOP_timeout: BLPOP on empty key with 1s timeout, verify timeout return
  results.push(await runTest('BLPOP_timeout', cat, async () => {
    const k = key(prefix, 'block:blpop_timeout');
    try {
      const start = Date.now();
      const result = await redis.blpop(k, 1);
      const elapsed = Date.now() - start;

      assertTrue(result === null, 'BLPOP should return null on timeout');
      assertTrue(elapsed >= 800 && elapsed <= 2500,
        `Timeout duration out of range: ${elapsed}ms`);
    } finally {
      await redis.del(k);
    }
  }));

  // BLPOP_multi_key: BLPOP on multiple keys, LPUSH to second key, verify correct key returned
  results.push(await runTest('BLPOP_multi_key', cat, async () => {
    const k1 = key(prefix, 'block:multi1');
    const k2 = key(prefix, 'block:multi2');
    try {
      await redis.rpush(k2, 'from_k2');
      const result = await redis.blpop(k1, k2, 1);
      assertEqual(k2, result[0]);
      assertEqual('from_k2', result[1]);
    } finally {
      await redis.del(k1, k2);
    }
  }));

  // BRPOP_basic: Same as BLPOP but from right side
  results.push(await runTest('BRPOP_basic', cat, async () => {
    const k = key(prefix, 'block:brpop');
    try {
      await redis.rpush(k, 'first', 'second');
      const result = await redis.brpop(k, 1);
      assertEqual(k, result[0]);
      assertEqual('second', result[1]);
    } finally {
      await redis.del(k);
    }
  }));

  // BLMOVE_basic: BLMOVE from source to dest list, verify move
  results.push(await runTest('BLMOVE_basic', cat, async () => {
    const src = key(prefix, 'block:blmove_src');
    const dst = key(prefix, 'block:blmove_dst');
    try {
      await redis.rpush(src, 'item1');
      const val = await redis.blmove(src, dst, 'LEFT', 'RIGHT', 1);
      assertEqual('item1', val);

      // Verify item is in destination
      const dstVals = await redis.lrange(dst, 0, -1);
      assertEqual(1, dstVals.length);
      assertEqual('item1', dstVals[0]);
    } finally {
      await redis.del(src, dst);
    }
  }));

  // BLMPOP_basic: BLMPOP with LEFT/RIGHT direction
  results.push(await runTest('BLMPOP_basic', cat, async () => {
    const k = key(prefix, 'block:blmpop');
    try {
      await redis.rpush(k, 'a', 'b', 'c');
      // BLMPOP timeout numkeys key ... LEFT|RIGHT
      const result = await redis.call('BLMPOP', 1, 1, k, 'LEFT');
      assertTrue(Array.isArray(result), 'BLMPOP should return an array');
      assertEqual(k, result[0]);
      assertTrue(Array.isArray(result[1]), 'BLMPOP second element should be an array');
      assertEqual('a', result[1][0]);
    } finally {
      await redis.del(k);
    }
  }));

  // BZPOPMIN_basic: BZPOPMIN with data and timeout variants
  results.push(await runTest('BZPOPMIN_basic', cat, async () => {
    const k = key(prefix, 'block:bzpopmin');
    try {
      await redis.zadd(k, 1, 'a', 2, 'b');
      const result = await redis.bzpopmin(k, 1);
      assertTrue(Array.isArray(result), 'BZPOPMIN should return an array');
      assertEqual(k, result[0]);
      assertEqual('a', result[1]);
      assertEqual('1', result[2]);
    } finally {
      await redis.del(k);
    }
  }));

  // BZPOPMAX_basic: BZPOPMAX with data and timeout variants
  results.push(await runTest('BZPOPMAX_basic', cat, async () => {
    const k = key(prefix, 'block:bzpopmax');
    try {
      await redis.zadd(k, 1, 'a', 2, 'b');
      const result = await redis.bzpopmax(k, 1);
      assertTrue(Array.isArray(result), 'BZPOPMAX should return an array');
      assertEqual(k, result[0]);
      assertEqual('b', result[1]);
      assertEqual('2', result[2]);
    } finally {
      await redis.del(k);
    }
  }));

  // BLOCKING_concurrent: Producer-consumer pattern
  results.push(await runTest('BLOCKING_concurrent', cat, async () => {
    const k = key(prefix, 'block:concurrent');
    const client2 = createClient();
    await client2.connect();

    try {
      // Schedule push after delay on second connection
      const pushPromise = new Promise((resolve, reject) => {
        setTimeout(async () => {
          try {
            await client2.rpush(k, 'async_val');
            resolve();
          } catch (e) {
            reject(e);
          }
        }, 300);
      });

      // BLPOP should block then receive pushed value
      const result = await redis.blpop(k, 5);
      assertEqual(k, result[0]);
      assertEqual('async_val', result[1]);

      await pushPromise;
    } finally {
      client2.disconnect();
      await redis.del(k);
    }
  }));

  // BLOCKING_timeout_accuracy: Verify timeout duration is within acceptable range
  results.push(await runTest('BLOCKING_timeout_accuracy', cat, async () => {
    const k = key(prefix, 'block:timeout_accuracy');
    try {
      const start = Date.now();
      const result = await redis.blpop(k, 1);
      const elapsed = Date.now() - start;

      assertTrue(result === null, 'BLPOP should return null on timeout');
      const diff = Math.abs(elapsed - 1000);
      assertTrue(diff <= 500,
        `Timeout accuracy out of range: expected ~1000ms, got ${elapsed}ms (diff: ${diff}ms)`);
    } finally {
      await redis.del(k);
    }
  }));

  return results;
});
